addons/sale/migrations/12.0.1.1/post-migration.py

```diff
@@ -19,16 +19,82 @@ def type_change_payment_transaction_and_sale_order(env):
         )
 
 
+def fill_sale_order_line_is_expense(cr):
+    # fill is_expense
+    openupgrade.logged_query(
+        cr, """
+        UPDATE sale_order_line sol
+        SET is_expense = TRUE
+        FROM account_analytic_line aal
+        JOIN product_product pp ON aal.product_id = pp.id
+        LEFT JOIN product_template pt ON pp.product_tmpl_id = pt.id
+        WHERE aal.so_line = sol.id AND aal.amount <= 0
+            AND pt.expense_policy IS NOT NULL AND pt.expense_policy != 'no'
+        """
+    )
+
+
+def fill_sale_order_line_qty_delivered_method(cr):
+    # set qty_delivered_method = 'analytic'
+    openupgrade.logged_query(
+        cr, """
+        UPDATE sale_order_line sol
+        SET qty_delivered_method = 'analytic'
+        WHERE is_expense = TRUE
+        """
+    )
+    if openupgrade.column_exists(cr, 'sale_order', 'warehouse_id'):
+        # sale_stock is installed
+        # set qty_delivered_method = 'stock_move'
+        openupgrade.logged_query(
+            cr, """
+            UPDATE sale_order_line sol
+            SET qty_delivered_method = 'stock_move'
+            FROM product_product pp
+            LEFT JOIN product_template pt ON pp.product_tmpl_id = pt.id
+            WHERE sol.product_id = pp.id AND sol.is_expense != TRUE
+                AND pt.type IN ('consu', 'product')
+            """
+        )
+    # set qty_delivered_method = 'timesheet'
+    openupgrade.logged_query(
+        cr, """
+        UPDATE sale_order_line sol
+        SET qty_delivered_method = 'timesheet'
+        FROM product_product pp
+        LEFT JOIN product_template pt ON pp.product_tmpl_id = pt.id
+        WHERE sol.product_id = pp.id AND sol.is_expense != TRUE
+            AND pt.type = 'service' AND pt.service_type = 'timesheet'
+        """
+    )
+    # set qty_delivered_method = 'manual'
+    openupgrade.logged_query(
+        cr, """
+        UPDATE sale_order_line sol
+        SET qty_delivered_method = 'manual'
+        WHERE qty_delivered_method IS NULL
+        """
+    )
+
+
 def recompute_sale_order_line_invoicing(env):
     sale_lines = env['sale.order.line'].search([])
     sale_lines._compute_untaxed_amount_invoiced()
     sale_lines._compute_untaxed_amount_to_invoice()
 
 
+def recompute_sale_order_line_qty_delivered(env):
+    sale_lines = env['sale.order.line'].search([])
+    sale_lines._compute_qty_delivered()
```

Comment:
With, you are possibly changing the old value, provoking that sales orders that were fully invoiced now are not. Possible cases:

- Manual changes in the field in previous versions.
- Modules that modifies the computation of this field.
- etc

And not only that, doing this on post avoids to have the proper computation according extra modules (for example, `sale_stock`).

We need to go over each of the SO lines that doesn't have manual method in end migration, call the compute method, and if the value is different, then put manual method and restore the previous value.

Reply:
done